File: app/src/main/java/com/vocaby/app/repositories/VocabyRepository.java
package com.vocaby.app.repositories;

import android.app.Application;

import com.vocaby.app.api.ApiManager;
import com.vocaby.app.api.VocabyApiService;
import com.vocaby.app.data.DataManager;
import com.vocaby.app.data.VocabyDatabase;
import com.vocaby.app.data.dao.VocabyDao;
import com.vocaby.app.data.entity.CustomDefinition;
import com.vocaby.app.data.entity.CustomEntry;
import com.vocaby.app.data.entity.CustomEntryGroup;
import com.vocaby.app.data.entity.Definition;
import com.vocaby.app.data.entity.EntryGroupWithDefinitions;
import com.vocaby.app.data.entity.EntryWithData;
import com.vocaby.app.data.entity.User;
import com.vocaby.app.data.entity.UserSaves;
import com.vocaby.app.data.entity.WordDefinitions;
import com.vocaby.app.models.DefinitionChanges;
import com.vocaby.app.models.DefinitionGroupModel;
import com.vocaby.app.models.DefinitionModel;
import com.vocaby.app.models.EntryModel;
import com.vocaby.app.models.GroupChanges;
import com.vocaby.app.models.EntryDataPackage;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class VocabyRepository {
    private final ApiManager apiManager;
    private final VocabyDao vocabyDao;
    private final DataManager dataManager;


    public VocabyRepository(Application application) {
        VocabyDatabase vocabyDatabase = VocabyDatabase.getDatabase(application);
        vocabyDao = vocabyDatabase.vocabyDao();
        apiManager = ApiManager.getInstance();
        dataManager = DataManager.getInstance(application);
    }

    // HISTORY
    public List<String> getHistory() {
        return dataManager.getHistory();
    }

    public List<String> writeHistory(String word) {
        return dataManager.writeHistory(word);
    }

    // Gets data from UI Thread
    public Single<List<String>> getDictionaryEntriesFromDB() {
        return vocabyDao.getDictionaryEntries()
                .subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<List<String>> populateDictionaryEntries() {
        return dataManager.populateDictionaryEntries();
    }

    public List<String> getDictionaryEntries() {
        return dataManager.getDictionaryEntries();
    }

    public Single<List<String>> getDictionaryEntriesByLetter(String letter) {
        return vocabyDao.getDictionaryEntriesByLetter(letter)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<EntryModel> getWordDataFromDatabase(String word) {
        return vocabyDao.getWordData(word)
                .map(this::covertToEntryModel)
                .onErrorReturnItem(new EntryModel(word))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<EntryModel> getWordDataFromDatabase(int id) {
        return vocabyDao.getWordData(id)
                .map(this::covertToEntryModel)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<EntryModel> getEntryData(int userId, String entry) {
        return vocabyDao.getUserEntryData(userId, entry)
                .map(this::convertEntryData)
                .onErrorReturnItem(new EntryModel(entry))
                .subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<EntryDataPackage> getWordDataPackageLocally(String word, int userId) {
        return Single.zip(
                getWordDataFromDatabase(word),
                getEntryData(userId, word),
                hasSave(word, userId),
                EntryDataPackage::new
        );
    }

    private EntryModel covertToEntryModel(WordDefinitions wordDefinitions) {
        EntryModel wordData = new EntryModel(wordDefinitions.word.getId(), wordDefinitions.word.getWord());
        if (wordDefinitions.word.getPronunciation() != null) {
            wordData.setPronunciation(wordDefinitions.word.getPronunciation());
        } else {
            wordData.setPronunciation("");
        }

        for (Definition data : wordDefinitions.definitions) {
            wordData.addDefinition(data.getPos(), data.getDefinition(), data.getSentence());
        }

        return wordData;
    }

    public Single<Integer> hasSave(String word, int id) {
        return vocabyDao.hasSave(word, id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<EntryModel> getRandomWord() {
        return vocabyDao.getRandomWord()
                .map(this::covertToEntryModel)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable addSave(UserSaves userSaves) {
        return vocabyDao.addSave(userSaves)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable removeSave(int id, String word) {
        return vocabyDao.removeSave(id, word)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<User> getUser(int id) {
        return vocabyDao.getCurrentUser(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<List<Long>> insertSavedWords(List<UserSaves> userSaves) {
        return vocabyDao.insertSavedWords(userSaves)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<List<String>> getUserSaves(int id) {
        // Main Thread
        return vocabyDao.getSaves(id)
                .subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable addUserSaves(List<UserSaves> saves) {
        return vocabyDao.addSaves(saves)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable removeUserSaves(List<String> saves) {
        return vocabyDao.removeSaves(saves)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<String> getRandomUserSave(int id) {
        return vocabyDao.getRandomSave(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<Long> createUser(User user) {
        return vocabyDao.createUser(user)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable deleteUser(User user) {
        return vocabyDao.removeUser(user)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable deleteAllUsers(int localId) {
        return vocabyDao.removeAllUsers(localId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public VocabyApiService getVocabyApiService(int type) {
        return apiManager.getVocabyApiService(type);
    }

    // CUSTOM USER ENTRIES
    public Completable deleteUserEntry(int entryId, String entry) {
        return vocabyDao.deleteUserEntry(entryId)
                .andThen(Completable.fromAction(() -> dataManager.deleteEntryFromDictionary(entry)))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable deleteUserEntryGroups(List<CustomEntryGroup> groups) {
        return vocabyDao.deleteCustomEntryGroups(groups)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable updateUserEntryGroups(List<CustomEntryGroup> groups) {
        return vocabyDao.updateCustomEntryGroups(groups)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<Integer> insertOrUpdateEntry(int userId, String entry, GroupChanges groupChanges, Map<String, DefinitionChanges> definitionChangesMap) {
        Single<Long> entryInsert = Single.just((long) groupChanges.getEntryId());

        if (groupChanges.getEntryId() == -1) {
            entryInsert = vocabyDao.insertCustomEntry(new CustomEntry(userId, entry, OffsetDateTime.now(ZoneOffset.UTC).toInstant().toEpochMilli()))
                    .flatMap(id -> Single.fromCallable(() -> {
                        dataManager.addEntryToDictionary(entry);
                        return id;
                    }));
        }

        return entryInsert
                .flatMap(id -> {
                    int entryId = id.intValue();
                    List<CustomEntryGroup> deletedGroups = new ArrayList<>();
                    for (DefinitionGroupModel group : groupChanges.getDeletedItems()) {
                        deletedGroups.add(new CustomEntryGroup(group.getGroupId()));
                    }

                    List<CustomEntryGroup> updatedGroups = new ArrayList<>();
                    for (DefinitionGroupModel group : groupChanges.getUpdatedItems()) {
                        updatedGroups.add(new CustomEntryGroup(
                                group.getGroupId(),
                                entryId,
                                group.getType(),
                                group.getOrder()
                        ));
                    }

                    List<CustomEntryGroup> addedGroups = new ArrayList<>();
                    for (DefinitionGroupModel group : groupChanges.getAddedItems()) {
                        addedGroups.add(new CustomEntryGroup(
                                entryId,
                                group.getType(),
                                group.getOrder()
                        ));
                    }

                    List<CustomDefinition> deletedDefinitions = new ArrayList<>();
                    for (DefinitionChanges definitionChanges : definitionChangesMap.values()) {
                        for (DefinitionModel definitionModel : definitionChanges.getDeletedItems()) {
                            deletedDefinitions.add(new CustomDefinition(
                                    definitionModel.getId(),
                                    definitionChanges.getGroupId(),
                                    definitionModel.getDefinition(),
                                    definitionModel.getExample(),
                                    definitionModel.getOrder()
                            ));
                        }
                    }

                    List<CustomDefinition> updatedDefinitions = new ArrayList<>();
                    for (DefinitionChanges definitionChanges : definitionChangesMap.values()) {
                        for (DefinitionModel definitionModel : definitionChanges.getUpdatedItems()) {
                            updatedDefinitions.add(new CustomDefinition(
                                    definitionModel.getId(),
                                    definitionChanges.getGroupId(),
                                    definitionModel.getDefinition(),
                                    definitionModel.getExample(),
                                    definitionModel.getOrder()
                            ));
                        }
                    }

                    return deleteUserEntryGroups(deletedGroups)
                            .andThen(updateUserEntryGroups(updatedGroups))
                            .andThen(insertUserEntryGroups(addedGroups))
                            .flatMapCompletable(ids -> {
                                List<DefinitionGroupModel> newGroups = groupChanges.getAddedItems();
                                for (int i = 0; i < newGroups.size(); i++) {
                                    DefinitionChanges definitionChanges = definitionChangesMap.get(newGroups.get(i).getType());
                                    if (definitionChanges != null)
                                        definitionChanges.setGroupId(ids.get(i).intValue());
                                }

                                List<CustomDefinition> addedDefinitions = new ArrayList<>();
                                for (DefinitionChanges definitionChanges : definitionChangesMap.values()) {
                                    for (DefinitionModel definitionModel : definitionChanges.getAddedItems()) {
                                        addedDefinitions.add(new CustomDefinition(
                                                definitionChanges.getGroupId(),
                                                definitionModel.getDefinition(),
                                                definitionModel.getExample(),
                                                definitionModel.getOrder())
                                        );
                                    }
                                }

                                return insertUserDefinitions(addedDefinitions);
                            }).andThen(updateUserDefinitions(updatedDefinitions))
                            .andThen(deleteUserDefinitions(deletedDefinitions))
                            .andThen(Single.just(entryId));
                });
    }

    public Single<Long> insertCustomEntry(int userId, String entry, long date) {
        return vocabyDao.insertCustomEntry(new CustomEntry(userId, entry, date))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<List<Long>> insertUserEntryGroups(List<CustomEntryGroup> groups) {
        return vocabyDao.insertCustomEntryGroups(groups)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable insertUserDefinitions(List<CustomDefinition> definitions) {
        return vocabyDao.insertCustomDefinitions(definitions)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable updateUserDefinitions(List<CustomDefinition> definitions) {
        return vocabyDao.updateCustomDefinitions(definitions)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Completable deleteUserDefinitions(List<CustomDefinition> definitions) {
        return vocabyDao.deleteCustomDefinitions(definitions)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<List<String>> getUserEntries(int userId) {
        return vocabyDao.getUserEntries(userId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    private EntryModel convertEntryData(EntryWithData data) {
        EntryModel entryData = new EntryModel(
                data.customEntry.getEntryId(),
                data.customEntry.getEntry()
        );

        List<DefinitionGroupModel> groups = new ArrayList<>();
        for (EntryGroupWithDefinitions group : data.groups) {
            DefinitionGroupModel groupModel = new DefinitionGroupModel(
                    group.entryGroup.getGroupId(),
                    group.entryGroup.getType(),
                    group.entryGroup.getOrder()
            );

            List<DefinitionModel> definitions = new ArrayList<>();
            for (CustomDefinition definitionData : group.definitions) {
                DefinitionModel definitionModel = new DefinitionModel(
                        definitionData.getDefinitionId(),
                        group.entryGroup.getType(),
                        definitionData.getDefinition(),
                        definitionData.getExample(),
                        definitionData.getOrder()
                );

                definitions.add(definitionModel);
            }

            Collections.sort(definitions);
            groupModel.setDefinitionData(definitions);
            groups.add(groupModel);
        }

        Collections.sort(groups);
        entryData.setDefinitionGroups(groups);

        return entryData;
    }

    // UTIL
    public Single<Integer> getSavesCount() {
        return vocabyDao.getSavesCount()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public Single<Integer> getUserCount() {
        return vocabyDao.getUserCount()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/com/vocaby/application/data/VocabyRepository.kt
package com.vocaby.application.data

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import androidx.preference.PreferenceManager
import com.google.gson.Gson
import com.vocaby.application.Constants
import com.vocaby.application.api.ApiManager
import com.vocaby.application.data.dao.VocabyDao
import com.vocaby.application.data.entity.*
import com.vocaby.application.exceptions.ApiException
import com.vocaby.application.exceptions.IllegalFileException
import com.vocaby.application.models.FeedbackModel
import com.vocaby.application.models.customentry.DefinitionChanges
import com.vocaby.application.models.customentry.GroupChanges
import com.vocaby.application.models.customentry.UserEntry
import com.vocaby.application.models.dictionary.*
import com.vocaby.application.models.profile.FaqModel
import com.vocaby.application.states.ValidState
import com.vocaby.application.utils.Formatter
import com.vocaby.application.utils.Logger
import java.io.BufferedWriter
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.OutputStreamWriter
import java.net.UnknownHostException
import java.util.*

class VocabyRepository(private val vocabyDao: VocabyDao, val application: Application) {
    private val userSharedPreference: SharedPreferences =
        application.getSharedPreferences(Constants.USER_ID_KEY, Context.MODE_PRIVATE)
    private val dictionaryCacheSharedPreferences: SharedPreferences =
        application.getSharedPreferences(Constants.DICTIONARY_CACHE_ID, Context.MODE_PRIVATE)
    private var userId: Int = userSharedPreference.getInt(Constants.CURRENT_USER_ID_KEY, 1)
    private val dataManager: DataManager = DataManager.getInstance(application)
    private val apiService  = ApiManager.apiService

    /** --------------------- ENTRY -------------------- **/
    suspend fun replaceEntry(original: EntryModel, remote: EntryModel): Int {
        vocabyDao.deleteEntry(Word(original.id))
        val id = vocabyDao.insertEntry(Word(remote.entry, remote.pronunciation, remote.lastUpdated)).toInt()
        val definitions = mutableListOf<Definition>()
        for(groupData in remote.definitionGroups) {
            for (definitionData in groupData.definitionData) {
                definitions.add(
                    Definition(
                    id,
                    definitionData.definition,
                    definitionData.example,
                    groupData.type)
                )
            }
        }

        vocabyDao.insertDefinitions(definitions)
        return id
    }

    suspend fun checkAndGetEntryDataFromApi(entry: String, date: String): EntryModel? {
        return try {
            val response = apiService.checkAndGetDefinitions(entry, date)

            val oldSet = dictionaryCacheSharedPreferences.getStringSet(Constants.SEARCH_CACHE, mutableSetOf<String>())!!
            val newSet = oldSet.toMutableSet()
            newSet.add(entry)
            dictionaryCacheSharedPreferences.edit().putStringSet(Constants.SEARCH_CACHE, newSet).apply()

            response.body()
        } catch (throwable: Throwable) {
            null
        }
    }

    fun checkApiCache(entry: String): Boolean {
        val set = dictionaryCacheSharedPreferences.getStringSet(Constants.SEARCH_CACHE, mutableSetOf<String>())!!
        return set.contains(entry)
    }

    fun clearDictionaryCache() {
        dictionaryCacheSharedPreferences.edit().clear().apply()
    }

    /** --------------------- USER -------------------- **/
    suspend fun setupUser() {
        val exists = vocabyDao.checkUser(userId)
        if (!exists) {
            val id = vocabyDao.createUser(User())
            userId = id.toInt()
            userSharedPreference.edit().putInt(Constants.CURRENT_USER_ID_KEY, userId).apply()
        }
    }

    /** --------------------- ENTRY -------------------- **/
    suspend fun getEntryDataFromDatabase(entry: String): EntryModel?
        = convertToEntryModel(vocabyDao.getEntryData(entry))

    private fun convertToEntryModel(wordDefinitions: WordDefinitions?): EntryModel? {
        wordDefinitions?.let {
            val pronunciation =
                wordDefinitions.wordData.pronunciation?.let { wordDefinitions.wordData.pronunciation }
                    ?: ""

            val wordData = EntryModel(
                wordDefinitions.wordData.id,
                wordDefinitions.wordData.word,
                pronunciation,
                wordDefinitions.wordData.lastUpdated
            )

            for (data in wordDefinitions.definitions) {
                wordData.addDefinition(data.pos, data.definition, data.sentence)
            }

            return wordData
        }

        return null
    }

    suspend fun getEntriesByCharacterFromDB(character: String): List<String> =
        vocabyDao.getDictionaryEntriesByCharacter(character)

    suspend fun getDailyPick(): DailyPick {
        val randomWordPicker = PreferenceManager.getDefaultSharedPreferences(application)
        val editor = randomWordPicker.edit()
        val lastTimeStarted = randomWordPicker.getInt(Constants.LAST_APP_STARTED, -1)
        val calendar = Calendar.getInstance()
        val today = calendar[Calendar.DAY_OF_YEAR]
        val dailyPick: DailyPick

        if (today != lastTimeStarted) {
            dailyPick = try {
                val response = apiService.getWoD(Formatter.formatDateToString(Date().time, precise=false))
                if (response.isSuccessful && response.code() == 200) {
                    DailyPick(response.body(), false)
                } else {
                    val data: WordDefinitions = vocabyDao.getRandomWord()
                    DailyPick(convertToEntryModel(data), true)
                }
            } catch (throwable: Throwable) {
                val data: WordDefinitions = vocabyDao.getRandomWord()
                DailyPick(convertToEntryModel(data), true)
            }

            dailyPick.entryModel?.let { it ->
                editor.putString(Constants.DICTIONARY_PICK_ID, it.entry)
                editor.putInt(Constants.LAST_APP_STARTED, today)
                editor.putBoolean(Constants.DICTIONARY_PICK_RANDOM, dailyPick.random)
                editor.apply()
            }

            return dailyPick
        } else {
            val pick = randomWordPicker.getString(Constants.DICTIONARY_PICK_ID, "vocaby")!!
            val random = randomWordPicker.getBoolean(Constants.DICTIONARY_PICK_RANDOM, true)
            val data: WordDefinitions? = vocabyDao.getEntryData(pick)
            return DailyPick(convertToEntryModel(data), random)
        }
    }

    suspend fun getAllEntryData(entry: String): EntryModel? {
        val customEntry = getUserEntryData(entry)
        customEntry?.let {
            return customEntry
        } ?: run {
            return getEntryDataFromDatabase(entry)
        }
    }

    /** --------------------- CUSTOM ENTRY -------------------- **/
    suspend fun getUserEntries(): LinkedList<UserEntry> {
        val list = vocabyDao.getUserEntries(userId)
        return LinkedList(list)
    }
    suspend fun getUserEntryData(entry: String): EntryModel? =
        convertCustomToEntryModel(vocabyDao.getUserEntryData(userId, entry))

    suspend fun removeCustomEntry(entryId: Int) = vocabyDao.deleteUserEntry(entryId)
    suspend fun removeCustomEntry(entry: String) = vocabyDao.deleteUserEntry(entry)
    suspend fun clearUserEntries() = vocabyDao.clearUserEntries(userId)
    suspend fun getAllUserEntries(): List<EntryModel> {
        val entries = vocabyDao.getAllUserEntryData(userId)
        val entryModels = mutableListOf<EntryModel>()
        entries.forEach { entryWithData ->
            val model = convertCustomToEntryModel(entryWithData)
            model?.let {
                entryModels.add(model)
            }
        }

        return entryModels
    }

    suspend fun insertOrUpdateEntry(
        entry: String,
        pronunciation: String,
        groupChanges: GroupChanges,
        definitionChangesMap: MutableMap<String, DefinitionChanges>,
        saveTime: Date
    ): Int {
        val entryId: Int = if (groupChanges.entryId == -1) {
            vocabyDao.insertCustomEntry(
                CustomEntry(
                    userId,
                    entry,
                    pronunciation,
                    saveTime
                )
            ).toInt()
        } else {
            vocabyDao.updateCustomEntry(
                CustomEntry(
                    userId,
                    entry,
                    pronunciation,
                    saveTime,
                    groupChanges.entryId
                )
            )

            groupChanges.entryId
        }

        val deletedGroups = mutableListOf<CustomEntryGroup>()
        for (group in groupChanges.deletedItems) {
            deletedGroups.add(CustomEntryGroup(group.groupId, group.type))
        }

        val updatedGroups = mutableListOf<CustomEntryGroup>()
        for (group in groupChanges.updatedItems) {
            updatedGroups.add(
                CustomEntryGroup(
                    group.groupId,
                    entryId,
                    group.type,
                    group.order
                )
            )
        }

        val addedGroups = mutableListOf<CustomEntryGroup>()
        for (group in groupChanges.addedItems) {
            addedGroups.add(
                CustomEntryGroup(
                    entryId,
                    group.type,
                    group.order
                )
            )
        }

        val deletedDefinitions = mutableListOf<CustomDefinition>()
        for (definitionChanges in definitionChangesMap.values) {
            for (definitionModel in definitionChanges.deletedItems) {
                deletedDefinitions.add(
                    CustomDefinition(
                        definitionModel.id,
                        definitionChanges.groupId,
                        definitionModel.definition,
                        definitionModel.example,
                        definitionModel.order
                    )
                )
            }
        }

        val updatedDefinitions = mutableListOf<CustomDefinition>()
        for (definitionChanges in definitionChangesMap.values) {
            for (definitionModel in definitionChanges.updatedItems) {
                updatedDefinitions.add(
                    CustomDefinition(
                        definitionModel.id,
                        definitionChanges.groupId,
                        definitionModel.definition,
                        definitionModel.example,
                        definitionModel.order
                    )
                )
            }
        }

        vocabyDao.deleteCustomEntryGroups(deletedGroups)
        vocabyDao.updateCustomEntryGroups(updatedGroups)

        val ids = vocabyDao.insertCustomEntryGroups(addedGroups)
        val newGroups = groupChanges.addedItems
        for (i in newGroups.indices) {
            val definitionChanges =
                definitionChangesMap[newGroups[i].type]
            if (definitionChanges != null) definitionChanges.groupId =
                ids[i].toInt()
        }

        val addedDefinitions = mutableListOf<CustomDefinition>()
        for (definitionChanges in definitionChangesMap.values) {
            for (definitionModel in definitionChanges.addedItems) {
                addedDefinitions.add(
                    CustomDefinition(
                        definitionChanges.groupId,
                        definitionModel.definition,
                        definitionModel.example,
                        definitionModel.order
                    )
                )
            }
        }

        vocabyDao.insertCustomDefinitions(addedDefinitions)
        vocabyDao.updateCustomDefinitions(updatedDefinitions)
        vocabyDao.deleteCustomDefinitions(deletedDefinitions)

        return entryId
    }

    suspend fun insertNewEntries(data: List<EntryModel>) {
        val customEntries = mutableListOf<CustomEntry>()
        for (entryData in data) {
            customEntries.add(CustomEntry(
                userId,
                entryData.entry,
                entryData.pronunciation,
                entryData.lastUpdated
            ))
        }

        val entryIds = vocabyDao.insertCustomEntries(customEntries)
        val addedGroups = mutableListOf<CustomEntryGroup>()
        for ((i, entryId) in entryIds.withIndex()) {
            for (group in data[i].definitionGroups) {
                addedGroups.add(
                    CustomEntryGroup(
                        entryId.toInt(),
                        group.type,
                        group.order
                    )
                )
            }
        }

        val groupIds = vocabyDao.insertCustomEntryGroups(addedGroups)
        val addedDefinitions = mutableListOf<CustomDefinition>()
        var j = 0
        for (entryModel in data) {
            for (group in entryModel.definitionGroups) {
                val definitionModels = group.definitionData
                for (definitionModel in definitionModels) {
                    addedDefinitions.add(
                        CustomDefinition(
                            groupIds[j].toInt(),
                            definitionModel.definition,
                            definitionModel.example,
                            definitionModel.order
                        )
                    )
                }
                j++
            }
        }

        vocabyDao.insertCustomDefinitions(addedDefinitions)
    }

    private fun convertCustomToEntryModel(data: EntryWithData?): EntryModel? {
        data?.let {
            val entryData = EntryModel(
                data.customEntry.entryId,
                data.customEntry.entry,
                data.customEntry.pronunciation,
                data.customEntry.lastUpdated
            )

            val groups: MutableList<DefinitionGroupModel> = ArrayList()
            for (group in data.groups) {
                val groupModel = DefinitionGroupModel(
                    group.entryGroup.groupId,
                    group.entryGroup.type,
                    group.entryGroup.order
                )

                val definitions: MutableList<DefinitionModel> = ArrayList()
                for (definitionData in group.definitions) {
                    val definitionModel = DefinitionModel(
                        definitionData.definitionId,
                        group.entryGroup.type,
                        definitionData.definition,
                        definitionData.example,
                        definitionData.order
                    )
                    definitions.add(definitionModel)
                }

                definitions.sort()
                groupModel.definitionData = definitions
                groups.add(groupModel)
            }

            groups.sort()
            entryData.definitionGroups = groups
            return entryData
        }

        return null
    }

    /** --------------------- SAVES -------------------- **/
    fun getSavedWordsFlow() = vocabyDao.getSavesFlow(userId)
    fun hasSaved(entry: String) = vocabyDao.hasSave(userId, entry)
    suspend fun getSavedWords() = vocabyDao.getSaves(userId)
    suspend fun addSaveItem(entry: String) {
        vocabyDao.addSave(UserSave(userId, entry))
    }

    suspend fun addSaveItems(saves: List<UserSave>) {
        vocabyDao.addSaves(saves)
    }

    suspend fun removeSaveItem(entry: String) = vocabyDao.removeSave(userId, entry)
    suspend fun clearSaves() = vocabyDao.clearSaves(userId)

    /** --------------------- IMPORT / EXPORT -------------------- **/
    fun importSavesFromExternalStorage(uri: Uri): List<UserSave> {
        val inputStream = application.contentResolver.openInputStream(uri)
        val reader = ObjectInputStream(inputStream)
        try {
            val saves = reader.readObject() as List<String>
            val userSaves = mutableListOf<UserSave>()
            for (save in saves) {
                userSaves.add(UserSave(userId, save))
            }

            return userSaves
        } catch (error: Throwable) {
            throw IllegalFileException(
                IllegalFileException.INVALID_FORMAT
            )
        } finally {
            inputStream?.close()
            reader.close()
        }
    }

    fun importEntriesBackupFromExternalStorage(uri: Uri): List<EntryModel> {
        val inputStream = application.contentResolver.openInputStream(uri)
        inputStream.use { ins ->
            ObjectInputStream(ins).use {
                return it.readObject() as List<EntryModel>
            }
        }
    }

    fun writeEntriesBackupToExternalStorage(entries: List<EntryModel>, uri: Uri) {
        application.contentResolver.openOutputStream(uri).use { outputStream ->
            val os = ObjectOutputStream(outputStream)
            os.writeObject(entries)
            os.flush()
            os.close()
        }
    }

    fun writeSavesBackupToExternalStorage(saves: List<String>, uri: Uri) {
        application.contentResolver.openOutputStream(uri).use { outputStream ->
            val os = ObjectOutputStream(outputStream)
            os.writeObject(saves)
            os.flush()
            os.close()
        }
    }

    fun writeSavesToExternalStorage(saves: List<String>, uri: Uri) {
        application.contentResolver.openOutputStream(uri).use { outputStream ->
            val bw = BufferedWriter(OutputStreamWriter(outputStream))
            val gson = Gson()
            gson.toJson(saves, bw)
            bw.flush()
            bw.close()
        }
    }

    fun writeEntriesToExternalStorage(entries: List<EntryModel>, uri: Uri) {
        application.contentResolver.openOutputStream(uri).use { outputStream ->
            val bw = BufferedWriter(OutputStreamWriter(outputStream))
            val gson = Gson()
            gson.toJson(entries, bw)
            bw.flush()
            bw.close()
        }
    }

    /** --------------------- HISTORY -------------------- **/
    fun writeToHistory(entry: SimpleEntryModel): List<SimpleEntryModel>? =
        dataManager.writeHistory(entry)

    fun getHistory(): LinkedList<SimpleEntryModel>? = dataManager.history
    fun clearHistory(): List<SimpleEntryModel>? = dataManager.clearHistory()

    /** --------------------- TYPES -------------------- **/
    suspend fun getTypes() = vocabyDao.getTypes()

    /** --------------------- DATA -------------------- **/
    suspend fun recordVisit(entryId: Int) {
        vocabyDao.recordVisit(DictionaryViewCount(0, userId, entryId, Formatter.formatDateToString(Date().time)))
    }

    suspend fun recordCustomVisit(entryId: Int) {
        vocabyDao.recordCustomVisit(CustomDictionaryViewCount(0, userId, entryId, Formatter.formatDateToString(Date().time)))
    }

    fun updateChartMode(displayAll: Boolean) {
        userSharedPreference.edit().putBoolean(Constants.CHART_MODE_ID, displayAll).apply()
    }

    fun getChartMode(): Boolean = userSharedPreference.getBoolean(Constants.CHART_MODE_ID, false)

    suspend fun getChartData(size: Int): List<VisitData> {
        val displayAll = userSharedPreference.getBoolean(Constants.CHART_MODE_ID, false)
        return if (displayAll) {
            vocabyDao.getAllSearchData(size)
        } else {
            vocabyDao.getMonthlySearchData(size)
        }
    }

    suspend fun eraseVisitData() {
        vocabyDao.deleteVisit(userId)
        vocabyDao.deleteCustomVisit(userId)
    }

    fun isUseConnectionEnabled() = userSharedPreference.getBoolean(Constants.CONNECTION_ID, true)

    fun setConnectionSettings(enabled: Boolean) {
        userSharedPreference.edit().putBoolean(Constants.CONNECTION_ID, enabled).apply()
    }

    fun isDataShareEnabled() = userSharedPreference.getBoolean(Constants.DATA_SHARE_ID, true)

    fun setDataShareSettings(enabled: Boolean) {
        userSharedPreference.edit().putBoolean(Constants.DATA_SHARE_ID, enabled).apply()
    }

    /** --------------------- FEEDBACK -------------------- **/
    suspend fun submitFeedback(feedbackModel: FeedbackModel): ValidState {
        return try {
            val response = apiService.submitFeedback("application/json", feedbackModel)
            if (response.isSuccessful) {
                ValidState.Valid
            } else {
                Logger.reportErrorToBugsnag(ApiException("${response.code()}: ${response.body().toString()}"))
                ValidState.Error("Vocaby's server is down :(")
            }
        } catch (e: UnknownHostException) {
            ValidState.Error("No internet connection")
        } catch (e: Throwable) {
            if (isDataShareEnabled()) Logger.reportErrorToBugsnag(e)
            ValidState.Error("Something went wrong...")
        }
    }

    /** --------------------- DATA -------------------- **/
    fun getFaq(): List<FaqModel> {
        return listOf(
            FaqModel(
                "Is Vocaby free?",
                "Yup! Vocaby is completely free and has no hidden fees or advertisements."
            ),
            FaqModel(
                "Why are some definitions outdated?",
                "Vocaby is powered by Princeton's Wordnet. " +
                        "At Vocaby, we are maintaining and updating definitions so that " +
                        "you are provided with the most up-to-date definition. " +
                        "If you would like to help improve the dictionary, please submit the form below."
            ),
            FaqModel(
                "Will definitions automatically update on my app?",
                "Yup! Once we make updates to our dictionary, your will retrieve the " +
                        "most up to date definitions on your app. This does require an " +
                        "internet connection though."
            ),
            FaqModel(
                "Does Vocaby collect data from me?",
                "We only collect error related data to improve the app and better your experience with Vocaby. " +
                        "If you don't feel comfortable sharing this data, you can opt out in the Data Management page."
            ),
            FaqModel(
                "If I do share my data, can it be traced back to me?",
                "No, the data does not contain any personally identifiable information that can trace back to you. " +
                        "The data does contain some information about your device but anything shared with us is securely encrypted."
            ),
            FaqModel(
                "Why does my import keep failing?",
                "Please make sure that your exported backup json " +
                        "file was indeed created by the app and was not tampered with. " +
                        "If you continue to experience this issue, please feel " +
                        "free to reach out to us!"
            ),
            FaqModel(
                "Will Vocaby be available on other platforms?",
                "We intend to increase Vocaby's " +
                        "availability across platforms further down the road, but we want " +
                        "to make sure that Vocaby matures on Android first."
            )
        )
    }
}